### DISCUSSION TITLE
Build Your own vs. Extending IDE

### DISCUSSION THREAD

**Discussion by srl295** [eclipse-theia/theia#14244] [Q&A] [answered]
Hi! I'm trying to evaluate the difference between the approaches in:

1. https://theia-ide.org/docs/composing_applications/
2. https://theia-ide.org/docs/blueprint_documentation/

It seems like `#1` is "running the Theia IDE, but with your own specific set of plugins",
where `#2` is "fork the Theia IDE (which is the top level gluing the components together) and customize it".

Do I have that right? It seems like `#2` is the right way to do branding, where `#1` has limited options for updating the brand.

I'm evaluating this for writing a custom editor for our application's data.

Thanks!

**Comment by msujew** [comment]
Hey @srl295,

Generally, both are simply a starting point from where you can create your own IDE. In theory, using `#2` you can get a custom branded fully-featured IDE in <30 minutes. It comes with a bunch of nice-to-haves (auto-updater, a sensible default plugin selection, dockerfile, electron build etc.) and runs out of the box.

However, a lot of adopters go with `#1`, because it offers the most flexibility from the start. It's not about using the "Theia IDE", but actually building your own IDE based on the Theia framework (this repository). That's how most adopters use Theia, see the [Arduino IDE](https://github.com/arduino/arduino-ide) or [Mbed Studio by ARM](https://os.mbed.com/studio/).

In the end, it doesn't matter too much which option you pick, since most of the features come from the underlying Theia framework anyway. Theia IDE just puts some nice-to-haves on top.

**Reply by srl295** [reply]
Concretely how does branding work with #1? It seems like the app itself is launched by the cli and so isn't as configurable other than the deps and plugins?

**Reply by msujew** [reply]
Adding onto Theia is not exclusively reserved to `@theia/*` packages. You can extend any behavior of Theia through dependency injection. See [this documentation section](https://theia-ide.org/docs/extensions/#theia-extensions). This is exactly the mechanism that the Theia IDE is using to provide all the additional features on top of "stock" Theia.

**Reply by srl295** [reply]
I think I'm still missing some steps from point A to B, but this is very helpful. I get that anything can be extended… 

I see that arduino launches the `@theia/cli`  in the [`./electron-app`](https://github.com/arduino/arduino-ide/tree/main/electron-app) directory, which has the application scaffolding (i'm assumign it's a parallel to the theia blueprint)… where does the `electron-app` directory come from? (seems to be a parallel to `applications/electron` in the blueprint). - containing the `*main.js` for the app.

**Reply by srl295** [reply]
Maybe `src-gen/backend/main.js` ?

**Reply by msujew** [reply]
I see, yeah the Theia CLI basically takes a look into the contributed dependencies (including your own) and figures out what to include into the app by looking at the `package.json`. The data in there is generated into the `src-gen` folder to then build a JavaScript artifact that includes any Theia extension.

**Reply by srl295** [reply]
OK. So if I wanted to do `#1` like arduino, how would I get started generating that `electron-app` or `applications/electron` directory?

1. is there a guide to building that?
2. do I copy from src-gen?
3. do I copy from the blueprint and customize it (which is sort of like `#2`)?
4. something else?

THanks!

**Reply by msujew** [reply, marked as answer]
I would recommend starting with the [yeoman generator](https://github.com/eclipse-theia/generator-theia-extension) (also linked [in here](https://theia-ide.org/docs/composing_applications#build-your-own-idetool)) which does all of that project setup for you. No need to dive into the inner workings of the CLI :)

**Reply by srl295** [reply]
@msujew it is linked there, but it says:

> This guide describes the manual steps to build a Theia-based product, there are two ways to avoid this manual set-up…

However, the guide that follows doesn't discuss those generated files- it implies that the guide is equivalent to running the generator..  Maybe a sentence contrasting the generator to the guide would help?

It could go at the very end of <https://theia-ide.org/docs/composing_applications>

**Reply by JonasHelming** [reply]
@srl295 Good feedback! Would you mind creating an issue [here](https://github.com/eclipse-theia/theia-website) and describing the mismatch or missing info a bit more in detail? Feedback like this is super important, I will try to address this!

**Reply by srl295** [reply]
It's a little related to  https://github.com/eclipse-theia/theia-website/issues/85 - do you want a new issue or comments there?

**Reply by JonasHelming** [reply]
New dedicated one would be better I think. Please note that we are about to deprecate/remove the manual json file approach in favor of the Yeoman generator as it is easier to maintain and more conveniant

**Reply by srl295** [reply]
https://github.com/eclipse-theia/theia-website/issues/611

**Reply by srl295** [reply]
@JonasHelming Next section, https://theia-ide.org/docs/blueprint_documentation/#application-icons how do i customize the icons when using the generator's generated app? (do you want an issue on the generator?)

**Reply by JonasHelming** [reply]
The documentation you are currently consuming is explicitly about customizing the Theia IDE, which can be used as a template for creating installable desktop applications based on Theia. The generator is about generating extensions along with an example product to host them. So most adopters get started with the generator and when they switch to building a product with branding and more, they fork the Theia IDE scripts and integrate their extension. I believe the real issue here might be that we obviously do not clarify what I write here clear enough. Does this make sense to you?

**Reply by srl295** [reply]
yup i think that's it and relevant to my questions above.  

I think I'm getting there, just figuring out all of the pieces…

**Reply by JonasHelming** [reply]
Please feel free to continue to ask questions. We are interested in improving the docs and materials we provide and also in supporting you in getting started.

**Comment by srl295** [comment]
Thanks I appreciate that, at this point I think I’m getting more
comfortable using the blueprint like a cookbook that is a source of ideas,
I guess it depends upon how you cook…

**Reply by srl295** [reply]
OK, some thoughts on going from the generator to the blueprint… There's a lot of stuff in the blueprint… For example, do I really want an updater that will update my app to the stock version? I think more documentation on how the blueprint was put together would be helpful , for example, it's not clear if I should add another extension to the extensions directory or add a new directory for my own extensions.

**Reply by JonasHelming** [reply]
So you mean a guide "From a simple app/extension to adopting Theia IDE as a blueprint"?

**Reply by JonasHelming** [reply]
See https://github.com/eclipse-theia/theia-website/issues/622
